File: product-brain-ts/src/backfill/pr-enrichment.ts
// Port target: ../product-brain/src/product_brain/backfill/pr_enrichment.py
import { execFileSync } from "node:child_process";
import type { Commit, PRComment, PullRequest } from "../models.js";

const PR_PAREN_RE = /\(#(\d+)\)/g;

function parseDt(s: unknown): Date | undefined {
  if (!s || typeof s !== "string") return undefined;
  const d = new Date(s.replace("Z", "+00:00"));
  return Number.isNaN(d.getTime()) ? undefined : d;
}

function gitRemoteRepo(repoPath: string): { owner: string; repo: string } | null {
  let url: string;
  try {
    url = execFileSync("git", ["-C", repoPath, "remote", "get-url", "origin"], {
      encoding: "utf8",
    }).trim();
  } catch {
    return null;
  }
  const m = /[:/]([^/:]+)\/([^/]+?)(?:\.git)?$/.exec(url);
  if (!m) return null;
  return { owner: m[1] ?? "", repo: m[2] ?? "" };
}

function prNumbersFromCommits(commits: Commit[]): number[] {
  const nums = new Set<number>();
  for (const c of commits) {
    const text = `${c.subject}\n${c.body ?? ""}`;
    PR_PAREN_RE.lastIndex = 0;
    let m: RegExpExecArray | null;
    while ((m = PR_PAREN_RE.exec(text)) !== null) {
      nums.add(Number(m[1]));
    }
  }
  return [...nums].sort((a, b) => a - b);
}

async function gh(url: string, token: string, params?: Record<string, string>): Promise<Response> {
  const u = new URL(url);
  if (params) for (const [k, v] of Object.entries(params)) u.searchParams.set(k, v);
  return fetch(u, {
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: "application/vnd.github+json",
    },
    signal: AbortSignal.timeout(30_000),
  });
}

export async function enrich(
  repoPath: string,
  ticketId: string,
  commits: Commit[],
  githubToken: string | undefined,
  enabled = true,
): Promise<PullRequest[]> {
  if (!enabled || !githubToken) return [];
  const remote = gitRemoteRepo(repoPath);
  if (!remote) return [];
  const { owner, repo } = remote;

  let prNumbers = prNumbersFromCommits(commits);
  if (prNumbers.length === 0) {
    const r = await gh("https://api.github.com/search/issues", githubToken, {
      q: `repo:${owner}/${repo} is:pr is:merged ${ticketId}`,
    });
    if (r.ok) {
      const data = (await r.json()) as { items?: Array<{ number: number }> };
      prNumbers = [...new Set((data.items ?? []).map((i) => i.number))].sort((a, b) => a - b);
    }
  }

  const out: PullRequest[] = [];
  for (const num of prNumbers) {
    const prResp = await gh(`https://api.github.com/repos/${owner}/${repo}/pulls/${num}`, githubToken);
    if (!prResp.ok) continue;
    const data = (await prResp.json()) as Record<string, unknown>;
    const commentsResp = await gh(
      `https://api.github.com/repos/${owner}/${repo}/pulls/${num}/comments`,
      githubToken,
      { per_page: "100" },
    );
    const reviewComments: PRComment[] = [];
    if (commentsResp.ok) {
      const items = (await commentsResp.json()) as Array<Record<string, unknown>>;
      for (const c of items) {
        const user = (c["user"] as Record<string, unknown> | undefined) ?? {};
        reviewComments.push({
          author: String(user["login"] ?? ""),
          body: String(c["body"] ?? ""),
          file: c["path"] as string | undefined,
          line: c["line"] as number | undefined,
          sha: c["commit_id"] as string | undefined,
        });
      }
    }
    const labels = ((data["labels"] as Array<Record<string, unknown>>) ?? []).map((l) => String(l["name"] ?? ""));
    out.push({
      number: num,
      title: String(data["title"] ?? ""),
      body: String(data["body"] ?? ""),
      labels,
      openedAt: parseDt(data["created_at"]),
      mergedAt: parseDt(data["merged_at"]),
      reviewComments,
    });
  }
  return out;
}

File: product-brain/src/product_brain/backfill/pr_enrichment.py
from __future__ import annotations

import re
import subprocess
from datetime import datetime
from pathlib import Path
from typing import Optional

import httpx

from ..models import Commit, PRComment, PullRequest


_PR_PAREN_RE = re.compile(r"\(#(\d+)\)")


def _parse_dt(s: Optional[str]) -> Optional[datetime]:
    if not s:
        return None
    return datetime.fromisoformat(s.replace("Z", "+00:00"))


def _git_remote_repo(repo_path: Path) -> Optional[tuple[str, str]]:
    try:
        r = subprocess.run(["git", "-C", str(repo_path), "remote", "get-url", "origin"],
                           capture_output=True, text=True, check=True)
    except subprocess.CalledProcessError:
        return None
    url = r.stdout.strip()
    m = re.search(r"[:/]([^/:]+)/([^/]+?)(?:\.git)?$", url)
    if not m:
        return None
    return m.group(1), m.group(2)


def _pr_numbers_from_commits(commits: list[Commit], ticket_id: str) -> list[int]:
    nums = set()
    for c in commits:
        for m in _PR_PAREN_RE.finditer(c.subject + "\n" + (c.body or "")):
            nums.add(int(m.group(1)))
    return sorted(nums)


def enrich(
    repo_path: Path,
    ticket_id: str,
    commits: list[Commit],
    github_token: Optional[str],
    enabled: bool = True,
) -> list[PullRequest]:
    if not enabled or not github_token:
        return []
    remote = _git_remote_repo(repo_path)
    if not remote:
        return []
    owner, repo = remote
    pr_numbers = _pr_numbers_from_commits(commits, ticket_id)

    headers = {
        "Authorization": f"Bearer {github_token}",
        "Accept": "application/vnd.github+json",
    }

    out: list[PullRequest] = []
    with httpx.Client(headers=headers, timeout=30) as client:
        if not pr_numbers:
            r = client.get(
                f"https://api.github.com/search/issues",
                params={"q": f"repo:{owner}/{repo} is:pr is:merged {ticket_id}"},
            )
            if r.status_code == 200:
                pr_numbers = sorted({i["number"] for i in r.json().get("items", [])})

        for num in pr_numbers:
            pr_resp = client.get(f"https://api.github.com/repos/{owner}/{repo}/pulls/{num}")
            if pr_resp.status_code != 200:
                continue
            data = pr_resp.json()
            comments_resp = client.get(
                f"https://api.github.com/repos/{owner}/{repo}/pulls/{num}/comments",
                params={"per_page": 100},
            )
            review_comments: list[PRComment] = []
            if comments_resp.status_code == 200:
                for c in comments_resp.json():
                    review_comments.append(PRComment(
                        author=c.get("user", {}).get("login", ""),
                        body=c.get("body", ""),
                        file=c.get("path"),
                        line=c.get("line"),
                        sha=c.get("commit_id"),
                    ))
            out.append(PullRequest(
                number=num,
                title=data.get("title", ""),
                body=data.get("body", "") or "",
                labels=[lbl.get("name", "") for lbl in data.get("labels", [])],
                opened_at=_parse_dt(data.get("created_at")),
                merged_at=_parse_dt(data.get("merged_at")),
                review_comments=review_comments,
            ))
    return out
